etl/helpers.py

```diff
@@ -263,3 +267,327 @@ def load_dependency(
+
+    return used_by
+
+
+def _recursive_get_all_dependencies_for_step_in_dag(
+    dag: Dict[str, Any], step: str, dependencies: Set[str] = set()
+) -> Set[str]:
+    if step in dag:
+        # If step is in the dag, gather all its substeps.
+        substeps = dag[step]
+        # Add substeps to the set of dependencies (union of sets, to avoid repetitions).
+        dependencies = dependencies | set(substeps)
+        for substep in substeps:
+            # For each of the substeps, repeat the process.
+            dependencies = dependencies | _recursive_get_all_dependencies_for_step_in_dag(
+                dag, step=substep, dependencies=dependencies
+            )
+    else:
+        # If step is not in the dag, return the default dependencies (which is an empty set).
+        pass
+
+    return dependencies
+
+
+def get_all_dependencies_for_step_in_dag(dag: Dict[str, Any], step: str) -> Set[str]:
+    """Get all dependencies for a given step in a dag.
+
+    This function returns all dependencies of a step, as well as their direct dependencies, and so on. In the end, the
+    result contains all datasets that the given step depends on, directly or indirectly.
+
+    Parameters
+    ----------
+    dag : Dict[str, Any]
+        Dag.
+    step : str
+        Step (as it appears in the dag).
+
+    Returns
+    -------
+    dependencies : Set[str]
+        All dependencies of a given step in a dag.
+    """
+    dependencies = _recursive_get_all_dependencies_for_step_in_dag(dag=dag, step=step)
+
+    return dependencies
+
+
+def get_all_usages_for_step_in_dag(dag: Dict[str, Any], step: str) -> Set[str]:
+    """Get all dependencies for a given step in a dag.
+
+    This function returns all datasets for which a given step is a dependency, as well as those datasets for which they
+    are also dependencies, and so on. In the end, the result contains all datasets that use, directly or indirectly, the
+    given step.
+
+    Parameters
+    ----------
+    dag : Dict[str, Any]
+        Dag.
+    step : str
+        Step (as it appears in the dag).
+
+    Returns
+    -------
+    dependencies : Set[str]
+        All usages of a given step in a dag.
+
+    """
+    # A simple solution is to simply reverse the graph, and apply the already existing function that finds all
+    # dependencies.
+    reverse_dag = reverse_graph(graph=dag)
+    dependencies = get_all_dependencies_for_step_in_dag(dag=reverse_dag, step=step)
+
+    return dependencies
+
+
+class ArchiveStepUsedByActiveStep(ExceptionFromDocstring):
+    """Archived steps have been found as dependencies of active steps.
+
+    The solution is either:
+    * To archive those active steps.
+    * To un-archive those archive steps.
+    """
+
+
+class LatestVersionOfStepShouldBeActive(ExceptionFromDocstring):
+    """The latest version of each data step should be in the dag as a main step (maybe it was accidentally removed)."""
+
+
+class VersionTracker:
```

Comment:
Is it used anywhere when we use special archive DAG file? If not, I'd be more inclined to merge only the archival-related part and move this to a different PR and wait for a moment when we actually need this. These methods are useful, but perhaps refactoring them and having them as [methods on DAG/Step classes](https://github.com/owid/etl/issues/801) might be better?

Reply:
I agree that some of the components of VersionTracker should be moved to a `DAG` class. I like the idea. But I think we still need a class that keeps those functions from DAG, plus additional ones (for example, it should also interact with grapher, to get the number of charts used by each dataset, or even user analytics about each dataset). The goal in the end is to have a tool that makes it easy to:
* Have an overview of our datasets, to detect which ones are outdated, and how to prioritize their updates
* Create/update/archive datasets.

So I'd merge it for now (it's not used anywhere else in the code anyway) and we can rethink its components in a separate PR. I will add a message in the docstring mentioning that it should not be used for now.